Add example of getting DataLoader from within LightningModule (#10903)
PyTorchLightning/pytorch-lightning: docs/source/guides/data.rst

--------------

********************************************
Accessing DataLoaders within LightningModule
********************************************

In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the ``Trainer`` object:

.. code-block:: python

    def test_step(self, batch, batch_idx, dataloader_idx):

>         test_dl = self.trainer.test_dataloaders[dataloader_idx]
>         test_dataset = test_dl.dataset

Reviewer: ```suggestion
        test_dl = self.trainer.test_dataloaders[dataloader_idx]
        test_dataset = test_dl.dataset
        test_sampler = test_dl.sampler
        ...
```

        # extract metadata, etc. from the dataset:
        ...

> Since Lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata.

Reviewer: ```suggestion
Since PyTorch Lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata.
```

Reviewer 2: I think just Lightning tells it's PyTorch Lightning. That's what we use at other places as well.

> If you are using :class:`~pytorch_lightning.trainer.supporters.CombinedLoader`, you can access the dataloaders using the following way:
> 
> If using a list of dataloaders:

Reviewer: ```suggestion
Evaluation dataloaders are iterated over sequentially. If you want to iterate over them in parallel, PyTorch Lightning provides a :class:`~pytorch_lightning.trainer.supporters.CombinedLoader` object which supports collection of dataloaders such as list, tuple, or dictionary. The dataloaders can be accessed using in the same way as the provided structure.:

If you provided a list of dataloaders:
```

.. code-block:: python

    from pytorch_lightning.trainer.supporters import CombinedLoader

    def test_dataloader(self):
        test_dl1 = ...
        test_dl2 = ...
        return CombinedLoader([test_dl1, test_dl2])

    def test_step(self, batch, batch_idx):

>         test_dl1 = self.trainer.test_dataloaders[0].loaders[0]

Reviewer: ```suggestion
        combined_loader = self.trainer.test_dataloaders[0]
        list_of_loaders = combined_loader.loaders
        test_dl1 = list_of_loaders.loaders[0]
```

> If using a dictionary of dataloaders:

Reviewer: ```suggestion
If you provided dictionary of dataloaders:
```

.. code-block:: python

    from pytorch_lightning.trainer.supporters import CombinedLoader

    def test_dataloader(self):
        test_dl1 = ...
        test_dl2 = ...
        return CombinedLoader({"dl1": test_dl1, "dl2": test_dl2})

    def test_step(self, batch, batch_idx):

>         test_dl1 = self.trainer.test_dataloaders[0].loaders["dl1"]

Reviewer: ```suggestion
        combined_loader = self.trainer.test_dataloaders[0]
        dictionarry_of_loaders = combined_loader.loaders
        test_dl1 = dictionarry_of_loaders["dl1"]
```

--------------

********************************************
Accessing DataLoaders within LightningModule
********************************************

In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the ``Trainer`` object:

> .. code-block:: python

Reviewer 3: and make it tested at least with syntax

> 

Reviewer 3: can we keep sense of `class` here
```suggestion
    class MyModule(LightningModule):
```

    def test_step(self, batch, batch_idx, dataloader_idx):
        test_dl = self.trainer.test_dataloaders[dataloader_idx]
        test_dataset = test_dl.dataset
        test_sampler = test_dl.sampler
        ...
        # extract metadata, etc. from the dataset:
        ...

Since Lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata.

> Evaluation dataloaders are iterated over sequentially. If you want to iterate over them in parallel, PyTorch Lightning provides a :class:`~pytorch_lightning.trainer.supporters.CombinedLoader` object which supports collection of dataloaders such as list, tuple, or dictionary. The dataloaders can be accessed using in the same way as the provided structure.:

Reviewer 4: ```suggestion
Evaluation dataloaders are iterated over sequentially. If you want to iterate over them in parallel, PyTorch Lightning provides a :class:`~pytorch_lightning.trainer.supporters.CombinedLoader` object which supports collections of dataloaders such as list, tuple, or dictionary. The dataloaders can be accessed using in the same way as the provided structure.:
```

If you provided a list of dataloaders:

.. code-block:: python

    from pytorch_lightning.trainer.supporters import CombinedLoader

    def test_dataloader(self):
        test_dl1 = ...
        test_dl2 = ...

>         return CombinedLoader([test_dl1, test_dl2])

Reviewer 5: The combined dataloaders are already explained above. Let's cut this short and not go into detail again here. This section should just be about accessing, not creating dataloaders.

    def test_step(self, batch, batch_idx):
        combined_loader = self.trainer.test_dataloaders[0]
        list_of_loaders = combined_loader.loaders
        test_dl1 = list_of_loaders.loaders[0]

If you provided dictionary of dataloaders:

.. code-block:: python

    from pytorch_lightning.trainer.supporters import CombinedLoader

    def test_dataloader(self):
        test_dl1 = ...
        test_dl2 = ...
        return CombinedLoader({"dl1": test_dl1, "dl2": test_dl2})

    def test_step(self, batch, batch_idx):
        combined_loader = self.trainer.test_dataloaders[0]

>         dictionarry_of_loaders = combined_loader.loaders
>         test_dl1 = dictionarry_of_loaders["dl1"]

Reviewer 4: ```suggestion
        dictionary_of_loaders = combined_loader.loaders
        test_dl1 = dictionary_of_loaders["dl1"]
```
Just a minor typo

    # Set limit_val_batches as an int
    trainer = Trainer(limit_val_batches=100)

> 

Reviewer 5: ```suggestion
----------

```

> ********************************************
> Accessing DataLoaders within LightningModules
> ********************************************

Reviewer 5: ```suggestion
*********************************************
Accessing DataLoaders within LightningModules
*********************************************
```

    # Set limit_val_batches as an int
    trainer = Trainer(limit_val_batches=100)

> *****************************************
> Accessing DataLoaders in LightningModules
> *****************************************

Reviewer 2: ```suggestion
********************************************
Accessing DataLoaders within LightningModule
********************************************
```

> In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the `Trainer` object:

Reviewer 2: ```suggestion
In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the ``Trainer`` object:
```

.. code-block:: python

>     def test_step(batch, batch_idx, dataloader_idx):

Reviewer 2: ```suggestion
    def test_step(self, batch, batch_idx, dataloader_idx):
```

>         test_dl = self.trainer.test_dataloaders[dataloader_idx]

Reviewer 2: I think we should add example when using `CombinedLoader` and one for train dataloader. Train dataloader is always a combined dataloader
When someone use `CombinedLoader`, the dataloaders are accessible using `self.test_dataloader.loaders`

Author: I'm not familiar with the internals of `CombinedLoader` -- if you don't mine providing an example I'd be happy to add it here

Reviewer 5: Here is an example:

```py
def training_step(self, batch, batch_idx):
        train_dl = self.trainer.train_dataloadesr.loaders[0]
```

Reviewer 2: same we can extend for eval dataloaders having dict return from eval_dataloader

```py
from pytorch_lightning.trainer.supporters import CombinedLoader

def val_dataloader(self):
    val_dl1 = ...
    val_dl2 = ...
    return CombinedLoader({'dl1: val_dl1, 'dl2': val_dl2})

def validation_step(self, batch, batch_idx):
    val_dl1 = self.val_dataloaders[0]['val_dl1'] 
```

Reviewer: It should be self.val_dataloaders.loaders['val_dl_1']

        test_dataset = test_dl.dataset
        # extract metadata, etc. from the dataset:
        ...

> Since pytorch-lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata

Reviewer 2: ```suggestion
Since Lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata.
```
